### ISSUE TITLE
Can I Integrating Cellpose Model into RectLabel Pro for Enhanced Cell Segmentation

### ISSUE THREAD

**Issue by zengzpi** [ryouchinsa/Rectlabel-support#269]
I am using RectLabel Pro to label cells as shown below:
<img width="945" alt="image" src="https://github.com/user-attachments/assets/090484e6-7137-4768-b547-51210b58bde4" />
However, I find that Cellpose is better suited for segmenting these objects due to its advanced accuracy and efficiency. I would like to integrate the Cellpose model into RectLabel Pro to automate and enhance my cell segmentation workflow.

I have converted the Cellpose model into an .mlpackage, but it seems I cannot load the Core ML model due to an input error.

Questions:

can I integrate the Cellpose model into RectLabel Pro to generate segmentation masks directly instead of SAM?

**Comment by ryouchinsa** [comment]
Thanks for writing the issue.

We will start reading the Cellpose repository and find out how to convert to the Core ML model. If we could convert, we will implement on RectLabel to decode the output layer to the pixels mask and polygon.
https://github.com/MouseLand/cellpose

**Comment by zengzpi** [comment]
Great! I’m very excited about the possibility of successfully integrating this into the app. Hopefully, the steps I’ve outlined maybe help a little:

### **Step 1: Load the Cyto3 Model**
Load the Cyto3 model using the Cellpose library as shown below:
[Run Cyto3 Model](https://github.com/MouseLand/cellpose/blob/main/notebooks/run_cyto3.ipynb)

```python
from cellpose import models
model = models.Cellpose(gpu=True, model_type="cyto3")
```

---

### **Step 2: Save the Model to a `.pt` File Using TorchScript**

```python
import torch

device = torch.device("cpu")  # Use "cpu" in order to successfully convert into mlpackage in a macos

# 1) Move the model to the target device (CPU or GPU)
model.cp.net.to(device)
model.cp.net.eval()

# Example input for tracing (adjust the shape as needed based on your model's input)
example_input = torch.rand(1, 2, 256, 256)

# Trace the model using TorchScript
traced_model = torch.jit.trace(model.cp.net, example_input)

# Save the traced model
traced_model.save("cellpose_model.pt")
```

---

### **Step 3: Convert to `.mlpackage` Using CoreMLTools**

```python
import coremltools as ct

# Load the TorchScript model
model = ct.convert(
    model="cellpose_model.pt", 
    source="pytorch",
    inputs=[ct.TensorType(name="input", shape=(1, 2, 256, 256))],
    minimum_deployment_target=ct.target.macOS15  # Specify the deployment target
)

# Optionally, specify input types for better conversion
# input_shape = (1, 3, 256, 256)  # Adjust based on your model's input if necessary
# model = ct.convert("cellpose_model.pt", inputs=[ct.TensorType(name="input", shape=input_shape)])

# Save the CoreML model
model.save("cellpose.mlpackage")
```

---

### **Notes:**
1. Step 1 and Step 2 maybe run in a linux system.
3. The CoreMLTools library requires macOS for creating and deploying `.mlpackage` files.

**Comment by ryouchinsa** [comment]
Thanks for writing the details.

To trace the PyTorch model correctly, Tensor should be used in CellPose code, but Numpy arrays are used for some modules, so that to convert to CoreML correctly, we have to rewrite CellPose code.

```
ts = torch.jit.trace(model, im, strict=False) 


/Users/ryo/Downloads/cellpose-main/cellpose/transforms.py:476: TracerWarning: Converting a tensor to a NumPy array might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  x = x.cpu().numpy()
```

**Comment by ryouchinsa** [comment]
The new version 2025.08.28 was released.
Please let us know your opinion.

- Added ONNX menu -> Load Cellpose ONNX Model.
- Added "Export a Cellpose 16bit PNG file" option when exporting mask images.
- Added "Import Cellpose 16bit PNG files" on Export menu.

# Load Cellpose ONNX Model
Read [Cellpose CPP Wrapper for macOS and Ubuntu GPU](https://github.com/ryouchinsa/cellpose-cpp).

Select a model.
- [cyto3.onnx](https://huggingface.co/rectlabel/cellpose/resolve/main/cyto3.onnx.zip)
- [cpsam.onnx](https://huggingface.co/rectlabel/cellpose/resolve/main/cpsam.onnx.zip)

Select a create type.
- polygon
- pixels

Create using Cellpose.
- The first channel is for the cytoplasm and the second channel is for the nuclear.
- Increase the diameter to detect larger cells, decrease the diameter to detect smaller cells.
- Increase the flow threshold to increase the number of detections.
- Increase the min size not to detect cells which are less than the min size.
- Press start button to start downloading the model and processing the image.

View menu.
- "Focus box" to quick zoom to the selected polygon/pixels.
- "Hide other boxes" to hide other polygons.
- "Hide pixels" to toggle pixels alpha.
- "Show small pixels" to show small pixels areas.
- "Show other pixels" to show other pixels objects.
- "Show indexes on labels table" to show each row number on the labels table.
- "Show checkboxes on labels table" to show a checkbox to show/hide for each row on the labels table.
- "Show depth image" to show the predicted Cellpose.

<video src="https://github.com/user-attachments/assets/22048c89-2412-43e3-9c0d-d0e12134ded5" controls="controls" muted="muted" class="width-fit" style="max-height:640px; min-height: 200px"></video>

**Comment by ryouchinsa** [comment]
The new version 2025.11.22 was released.

Improved the Cellpose feature for gray images, for gray images, for channels, please set the same one such as (Red, Red) for better detections. 

We solved the problem for gray images.
For gray images, when specified the channels [0, 0], we should set gray pixel values for the first channel, and for the second channel, should set 0.
In the previous version, for both channels, we were setting the same gray pixel values.
This caused the accuracy to be decreaesd.

Using the ONNX model, the image is resized to (512, 512), but cellpose-gui does not have to resize the image.
For larger images and the same diameter 30, cellpose-gui tends to have more detections.
For larger images, using smaller diameter such as 20, cellpose-cpp can have more detections.
If we compare using the (512, 512) resized image, cellpose-cpp and cellpose-gui predict almost the same result.

RectLabel

![Image](https://github.com/user-attachments/assets/0bbd9b15-09a7-4429-a69d-e3c63271e65f)

Cellpose-gui

![Image](https://github.com/user-attachments/assets/eeddf03f-f826-4b4a-b9a9-543da3cb4354)
